### DISCUSSION TITLE
How to attach a session to outer transaction when engine has isolation_level equal to AUTOCOMMIT?

### DISCUSSION THREAD

**Discussion by albertalexandrov** [sqlalchemy/sqlalchemy#11836] [Usage Questions] [answered]
Hi!

I'd like to implement commiting flow as it is in Django. I mean by default every single sql request is being autocommited and if one need to execute sql queries in transaction then create transaction and make session commits into that transaction. For that I developed some code:

```python
from typing import Annotated

import sqlalchemy as sa
import uvicorn
from fastapi import FastAPI, Depends
from pydantic import BaseModel, Field
from sqlalchemy import MetaData
from sqlalchemy.ext.asyncio import AsyncSession, async_sessionmaker, create_async_engine
from sqlalchemy.orm import declarative_base

app = FastAPI()

metadata = MetaData()
Base = declarative_base(metadata=metadata)
engine = create_async_engine("postgresql+asyncpg://postgres:postgres@localhost:5433/portal-podrjadchika-local", isolation_level="AUTOCOMMIT", echo=True)
SessionLocal = async_sessionmaker(bind=engine, expire_on_commit=False)


class User(Base):
    __tablename__ = "users"

    id = sa.Column(sa.Integer, autoincrement=True, primary_key=True, index=True)
    first_name = sa.Column(sa.String)
    last_name = sa.Column(sa.String)


class InputSchema(BaseModel):
    first_name: Annotated[str, Field(example="Tony")]
    last_name: Annotated[str, Field(example="Stark")]


async def get_session() -> AsyncSession:
    connection = await engine.connect()
    transaction = await connection.begin()
    session = AsyncSession(bind=connection)
    try:
        yield session
        await transaction.commit()
    except Exception as e:
        await transaction.rollback()  # does not rollback
        raise e
    finally:
        await session.close()
        await transaction.close()
        await connection.close()


@app.on_event("startup")
async def on_startup() -> None:
    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.drop_all)
        await conn.run_sync(Base.metadata.create_all)


async def create_user(session, data):
    user = User(**data)
    session.add(user)
    await session.commit()
    return user


@app.post('/users')
async def get_users(data: InputSchema, session: AsyncSession = Depends(get_session)):
    user = await create_user(session, data.model_dump())
    raise ValueError("Emulate exception after which all committed changes should be rollbacked")
    return user


if __name__ == "__main__":
    uvicorn.run("so:app", reload=True)
```

With `isolation_level` equal to AUTOCOMMIT rollback is not working. And I cannot understand why. 

Please help.

**Comment by zzzeek** [comment, marked as answer]
> I mean by default every single sql request is being autocommited and if one need to execute sql queries in transaction then create transaction and make session commits into that transaction.

this is how SQLAlchemy worked prior to version 2.0 , and this mode of operation has been removed.  See https://docs.sqlalchemy.org/en/20/changelog/migration_20.html#library-level-but-not-driver-level-autocommit-removed-from-both-core-and-orm 

I would strongly recommend using the "always transactional" of operation as it leads to much cleaner, more efficient and predictable code that works with SQLAlchemy's best practices and assumptions.

>With isolation_level equal to AUTOCOMMIT rollback is not working. 

if you set isolation level at AUTOCOMMIT then the DBAPI has no transactional capabilities at all.    There is no DBAPI autobegin therefore rollback and commit are no-ops.

If you want to mix DBAPI level AUTOCOMMIT and transactional engines, the easiest approach is to have two "option" engines (these share the same connection pool) where one uses AUTOCOMMIT and the other doesn't - see the second code example at https://docs.sqlalchemy.org/en/20/orm/session_transaction.html#setting-isolation-for-a-sessionmaker-engine-wide

**Reply by albertalexandrov** [reply]
Hmm. Got it. Thanks a lot for quick answer.
